### PR TITLE
Pool more data while pattern matching.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Tagging @ManishJayaswal 

This PR is purely about allocating less during expensive operations like 'NavigateTo'.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @grabalon

File: src/Features/Core/Portable/SpellCheck/AbstractSpellCheckCodeFixProvider.cs [hunk 1]
```diff
@@ -76,27 +76,44 @@ internal abstract class AbstractSpellCheckCodeFixProvider<TSimpleName> : CodeFix
...
+                    context, nameNode, nameText,
+                    completionList, similarityChecker).ConfigureAwait(false);
+            }
+            finally
+            {
+                similarityChecker.Free();
+            }
+        }
+
+        private async Task CheckItems(
```
Comment (Pilchie) [inline src/Features/Core/Portable/SpellCheck/AbstractSpellCheckCodeFixProvider.cs]: Should be CheckItemsAsync

File: src/Workspaces/Core/Portable/FindSymbols/SearchQuery.cs [hunk 2]
```diff
@@ -36,7 +36,7 @@ private SearchQuery(string name, SearchKind kind)
                     // its 'AreSimilar' method. That way we only create the WordSimilarityChecker
                     // once and it can cache all the information it needs while it does the AreSimilar
                     // check against all the possible candidates.
-                    var editDistance = new WordSimilarityChecker(name, substringsAreSimilar: false);
+                    var editDistance = WordSimilarityChecker.Allocate(name, substringsAreSimilar: false);
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/SearchQuery.cs]: You don't `Free` this?  Is that okay?  Should you comment?

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs [hunk 3]
```diff
@@ -351,7 +358,7 @@ private static bool ContainsUpperCaseLetter(string pattern)
                     }
 
                     var wordSpans = GetWordSpans(candidate);
-                    for (int i = 0; i < wordSpans.Count; i++)
+                    for (int i = 0, n = wordSpans.Count; i < n; i++)
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs]: Does this optimization actually help?

File: src/Workspaces/Core/Portable/Utilities/SpellChecker.cs [hunk 4]
```diff
@@ -97,68 +96,85 @@ public CacheResult(string candidate, bool areSimilar, double similarityWeight)
...
         }
 
-        public void Dispose()
+        public void Free()
         {
             _editDistance?.Dispose();
+            _source = null;
             _editDistance = null;
+            _lastAreSimilarResult = default(CacheResult);
+            s_pool.Push(this);
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Utilities/SpellChecker.cs]: Is it worth capping the pool size at all?

File: src/Workspaces/Core/Portable/Utilities/SpellChecker.cs [hunk 5]
```diff
@@ -97,68 +96,85 @@ public CacheResult(string candidate, bool areSimilar, double similarityWeight)
...
-        
+
         /// <summary>
         /// Returns true if 'originalText' and 'candidateText' are likely a misspelling of each other.
         /// Returns false otherwise.  If it is a likely misspelling a similarityWeight is provided
         /// to help rank the match.  Lower costs mean it was a better match.
         /// </summary>
         public static bool AreSimilar(string originalText, string candidateText, bool substringsAreSimilar, out double similarityWeight)
         {
-            using (var checker = new WordSimilarityChecker(originalText, substringsAreSimilar))
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Utilities/SpellChecker.cs]: You could leave it IDisposable and have the dispose put it back in the pool.

**Comment by CyrusNajmabadi** [comment]
> I'm sort of meh on this change overall - don't see the reason for removing IDisposable from WordSimilarityChecker if you still need to call Free on it.

I'm not a fan of "Dispose" adding something back to a pool.  The pattern we've gone with in the compiler is to have a "Free" method.  I like that that makes it very explicit.

File: src/Workspaces/Core/Portable/FindSymbols/SearchQuery.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SearchQuery.cs]: I'll comment.  It's not great, but it's non trivial to expand out disposal to the SearchQuery type as well.

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs]: Yes.  StringBreaks.Count has to actually search the embedded int to find out how many breaks it actaully has.  I think i'd like to make it a method to make it clear that it's not O(1).

File: src/Workspaces/Core/Portable/Utilities/SpellChecker.cs (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Utilities/SpellChecker.cs]: See below... i struggled with that idea, but in the end went with teh compiler pattern which i think makes it more clean that the object *is actually being returned*.

File: src/Features/Core/Portable/SpellCheck/AbstractSpellCheckCodeFixProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SpellCheck/AbstractSpellCheckCodeFixProvider.cs]: sure.

File: src/Workspaces/Core/Portable/FindSymbols/SearchQuery.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SearchQuery.cs]: Actually, i made SearchQuery disposable, and we now release that instance.

**Comment by sharwell** [comment]
> I'm not a fan of "Dispose" adding something back to a pool. The pattern we've gone with in the compiler is to have a "Free" method. I like that that makes it very explicit.

Basically what it comes down to is how much you care whether or not the object is returned to the pool. If it simply doesn't matter one way or the other, the approach used here is fine. However, if you would look at uses in a proposed change with the intent of verifying `Free` was called before objects go out of scope and request changes if someone forgot to do so, leave it as `IDisposable`.

**Comment by DustinCampbell** [comment]
>>I'm not a fan of "Dispose" adding something back to a pool. The pattern we've gone with in the compiler is to have a "Free" method. I like that that makes it very explicit.

>Basically what it comes down to is how much you care whether or not the object is returned to the pool. If it simply doesn't matter one way or the other, the approach used here is fine. However, if you would look at uses in a proposed change with the intent of verifying Free was called before objects go out of scope and request changes if someone forgot to do so, leave it as IDisposable.

IIRC, our good friend Mr. Lippert wrote a long email about why we shouldn't use Dispose in our ObjectPool classes.

**Comment by CyrusNajmabadi** [comment]
tagging @MattGertz This lowers the allocations we cause when typing and also using NavigateTo.

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by MattGertz** [comment]
Approved pending tests.

**Comment by CyrusNajmabadi** [comment]
Tagging @JoshVarty

**Pull request merged**